### PR TITLE
Add iexamine actors, move iexamine::cardreader to actor

### PR DESCRIPTION
**Pull request by anothersimulacrum**
#### Summary
SUMMARY: Infrastructure "Allow adding data for examine actions in JSON" 

#### Purpose of change
Terrain/furniture examination has really been hurting for a way to inject some data in JSON into what it does, so here it is.

#### Describe the solution
Add iexamine actors, modeled after iuse/mattack actors.
Replace the hardcoded ardreader examine function(s) with an actor to demonstrate use.

##### Commits
**Add iexamine actors, replace cardreader actor** - b36fda8

Terrain/furniture examination has long need a system where, instead of a hardcoded action, the changes are performed based on data specified in the terrain entry. This adds that system.

Add a virtual iexamine_actor, from which subclasses inherit, depending on the type. This is in large part based on the iuse_actor infrastructure.
This contains a string informing what type the actor is, for querying whether a terrain or furniture has this actor and for determining what subactor to load.

As an example here, the cardreader iexamine function is converted to be an examine actor.

**Move examine actions to separate doc** - e2bee0b

I'm going to be expanding documentation on these functions in the future, so move it to a separate doc to make it more clear/less messy.

**Document cardreader iexamine actor** - b95ecfd 

Expand the iexamine doc and document the cardreader actor in specific.

**Add update mapgen support for cardreader examine** - 57b1250

Move the application of the cardreader examine to a separate function, and in there distinguish between mapgen_update and the terrain/furniture changes.

#### Testing
You can open lab doors with science cards.
Adding a furniture change swapping `f_null` to something, testing the furniture changes.
Replace the terrain changes with a `mapgen_id` of `microlab_shifting_hall` for any cardreader, watch chaos ensue.

#### Additional context
I'm punting off fully integrating hacking to down the road. Right now it requires that all cardreaders it works with behave identically to the previous function.
Some more things could probably be using this actor, but those may need additional support/consideration.

### CODE REVIEW

File: src/iexamine_actors.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
+#include "iexamine.h"
+
```
Comment (BrettDong) [inline src/iexamine_actors.h:2]: Probably can resolve GCC 7 build error:
```suggestion
#include <memory>
```

**Comment by BrettDong** [comment]
Oh no I commented on the wrong place. Adding `#include <memory>` in `iexamine.h` should solve the problem.

**Comment by Night-Pryanik** [comment]
Does `terrain_changes` accepts an array of objects?

**Comment by anothersimulacrum** [comment]
It's an an object with multiple fields, e.g.
```json
{
  "t_1": "t_11",
  "t_2": "t_12",
  "t_3": "t_13"
}
```
`furn_changes` is the same. I will update the documentation to clarify.

**Comment by Night-Pryanik** [comment]
Good. It would allow easily create new pairs of "openable-by-cardreader" terrains.

File: src/mapdata.cpp [hunk 2]
```diff
@@ -408,14 +409,23 @@ bool map_data_common_t::has_examine( iexamine_function_ref func ) const
     return examine_func == &func;
 }
 
+bool map_data_common_t::has_examine( const std::string &action) const
```
Comment (kevingranade) [inline src/mapdata.cpp:412]:
```suggestion
bool map_data_common_t::has_examine( const std::string &action ) const
```

**Comment by Ramza13** [comment]
This is awesome.  I'm still looking it over but after the release I hopefully will integrate some of this with effect_on_conditions to allow all kinds of arbitrary examine actions defined in json.

**Pull request merged**